linera-ethereum/src/provider.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use alloy::{
+    primitives::{Address, U256},
+    providers::{Provider, ProviderBuilder, RootProvider},
+    rpc::types::eth::{
+        request::{TransactionInput, TransactionRequest},
+        Filter,
+    },
+    transports::http::reqwest::Client,
+};
+use alloy_primitives::Bytes;
+use async_trait::async_trait;
+use reqwest::Client as Client_json_ser;
+use url::Url;
+
+use crate::client::{EthereumQueries, JsonRpcClient};
+
+pub type HttpProvider = RootProvider<alloy::transports::http::Http<Client>>;
+
+use crate::{
+    client::get_block_id,
+    common::{event_name_from_expanded, parse_log, EthereumEvent, EthereumServiceError},
+};
+
+/// The Ethereum endpoint and its provider used for accessing the ethereum node.
+pub struct EthereumClientSimplified {
+    pub url: String,
+}
+
+#[async_trait]
+impl JsonRpcClient for EthereumClientSimplified {
+    type Error = EthereumServiceError;
+    async fn request_inner(&self, payload: Vec<u8>) -> Result<Vec<u8>, Self::Error> {
+        let client = Client_json_ser::new();
+        let res = client
+            .post(self.url.clone())
+            .json_ser(payload)
+            .send()
+            .await?;
+        let body = res.bytes().await?;
+        Ok(body.as_ref().to_vec())
+    }
+}
+
+impl EthereumClientSimplified {
+    /// Connects to an existing Ethereum node and creates an `EthereumEndpoint`
+    /// if successful.
+    pub fn new(url: String) -> Self {
+        Self { url }
+    }
+}
+
+pub struct EthereumClient<M> {
+    pub provider: M,
+}
+
+#[async_trait]
+impl EthereumQueries for EthereumClient<HttpProvider> {
```

Comment:
Why are there two `impl EthereumQueries` now, and neither of them can be used inside a smart contract?

Will we implement `Transport` inside the contract, and use `RootProvider`? Or something like `EthereumClient<SomethingElse>`?

Reply:
There are two implementation because of internal checks. But I can remove the old one.

We need to have the provider being used. When doing some operations on the Ethereum chain, we cannot support everything by the `JsonRpcClient` I think.

Yes, neither of them can be used in a smart contract, but the implementation should be relatively easy from now: Simply modify the code in `linera-sdk` to either take a pair `(URL, payload)` or a serialization of both.